Changes to support cross-partition querying with joined associations

webandtech opened the pull request:
This PR allows the query rewriter to be active when making queries without a current tenant and augments any joins to ensure that multi-tenant tables joined on to the query will reference the partition key from the source table, allowing the joins to be pushed down even when querying across multiple tenants.

This allows support for a query such as: `User.eager_load(:features).where(tenant_id: [a,b])`, which will now be pushed down to the 2 shards instead of causing a postgres error.

In addition, we can now support: `User.includes(:features).references(:features).where('features.new = true').distinct.count`. This is a COUNT(DISTINCT) that isn't allowed by citus across partitions but can be rewritten with a subquery in the query rewriter

List of changes:

* Query rewriter is now active when `MultiTenant.current_tenant` is `nil`
* Adds the partition key to join nodes based on the query's SELECT FROM source table instead of always adding the MultiTenant.current_tenant_id, which allows pushing down joins for multiple tenants in one query
* Ensures any `SELECT foo, bar DISTINCT` is rewritten as `SELECT foo, bar GROUP BY foo, bar` per the SQL workarounds here: https://docs.citusdata.com/en/v6.2/reference/sql_workarounds.html#select-distinct
* By default, also rewrite `SELECT COUNT(DISTINCT users.id)` as `SELECT COUNT(*) FROM (SELECT ... GROUP BY users.id, users.partition_key) subquery`
  * Allow this behavior to be disabled with `MultiTenant.with_hll_counts` to use approximate counts with the HLL extension

lfittl commented:
@webandtech Nice, thanks for rebasing that!

Any idea on the test failures on Rails 4.0 and 4.1? Looks like `arel/visitors/reduce` doesn't exist there.

webandtech commented:
@lfittl I moved the subquery Arel extension into query_rewriter.rb and added handling for Rails < 4.2 but it looks like there are still issues... perhaps you can take a look when you have a chance

webandtech commented:
We've changed our strategy to not do eager loads, which has made these changes unnecessary for us. Perhaps us or someone else will take it back up down the line.

lfittl commented:
@webandtech Gotcha, makes sense.

Thanks for pushing this up as you worked on it, its definitely helpful as a reference for future work in this area.

DaveWhitmer commented:
+1 to add this functionality
